app/controllers/reservations_controller.rb

```diff
@@ -0,0 +1,53 @@
+class ReservationsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :require_donor
+
+
+  def ship
+    @reservation = Reservation.find(params[:id])
+
+    if @reservation.donor_id == current_user.id
+      @reservation.ship!
+      render json: @reservation
+    else
+      redirect_to root_path, alert: "You are not authorized to access that page."
+    end
+  end
+
+  # TODO - should donors be able to mark their own gifts as received?
+  def receive
+    @reservation = Reservation.find(params[:id])
+
+    if @reservation.donor_id == current_user.id
+      @reservation.receive!
+      render json: @reservation
+    else
+      redirect_to root_path, alert: "You are not authorized to access that page."
+    end
+  end
+
+  def destroy
```

Comment:
reservations shouldn't be destroyable unless they're in the `reserved` or `shipped` stages.